### PR TITLE
Quiz Shortcode: Offer Ability to Display Quizzes Together

### PR DESCRIPTION
**Pull request by Aurorum**
#### Changes proposed in this Pull Request:

* This would allow people to display quizzes as one instead of them being listed individually throughout the post by wrapping the shortcode in `[quiz-wrapper]`

Screenshot 2021-08-06 at 17 17 38

#### Jetpack product discussion
None, but I don't imagine this will be a particularly controversial change. In my view, it's a significant enhancement to the existing Quiz system, and it's something I've hoped would be implemented for quite a while. Currently, all the quizzes are listed individually, and although that option will remain, it can quickly clutter a post.

#### Does this pull request change what data or activity we track or use?

Nope.

#### Testing instructions:

With Shortcodes enabled, try this. You should be able to now click through the Quiz, as it should all be displayed together. One of the reasons I've wanted this is to try something like Who Wants to be a Millionaire on my own site - this would allow something more like that compared to the existing Quiz system.

```
[quiz-wrapper]

[quiz]
[question]£100: Complete the famous phrase - "the writing is on the..."
[/question]

[answer]Wall[/answer]
[wrong]Notebook[/wrong]
[wrong]Building[/wrong]
[wrong]Window[/wrong]
[/quiz]

[quiz]
[question]£200: On a computer, "ALT" is short for?
[/question]

[answer]Alternate[/answer]
[wrong]Altitude[/wrong]
[wrong]Altogether[/wrong]
[wrong]Alter[/wrong]
[/quiz]

[quiz]
[question]£300: When drawing straws, a person gets chosen by picking which straw?
[/question]

[answer]Shortest straw[/answer]
[wrong]Last straw[/wrong]
[wrong]Thickest straw[/wrong]
[wrong]Bent straw[/wrong]
[/quiz]

[quiz]
[question]£500: William Shakespeare included the line "To be or not to be - that is the question" in which play?
[/question]

[answer]Hamlet[/answer]
[wrong]The Merchant of Venice[/wrong]
[wrong]Macbeth[/wrong]
[wrong]Othello[/wrong]
[/quiz]

[quiz]
[question]£1000: In the Chronicles of Narnia series, what is the name of the lion who appears?
[/question]

[answer]Aslan[/answer]
[wrong]Balrus[/wrong]
[wrong]Dagen[/wrong]
[wrong]Igor[/wrong]
[/quiz]

[quiz]
[question]£2000: What is the name for common table sugar?
[/question]

[answer]Sucrose[/answer]
[wrong]Glucose[/wrong]
[wrong]Fructose[/wrong]
[wrong]Maltose[/wrong]
[/quiz]

[quiz]
[question]£4000: Cleopatra was said to have been killed by what?
[/question]

[answer]Snake[/answer]
[wrong]Strangulation[/wrong]
[wrong]Knife[/wrong]
[wrong]Tripping[/wrong]
[/quiz]

[quiz]
[question]£8000: How many stars are there on the Australian flag?
[/question]

[answer]6[/answer]
[wrong]7[/wrong]
[wrong]8[/wrong]
[wrong]9[/wrong]
[/quiz]

[quiz]
[question]£16,000: The Ghats mountains are located in what country?
[/question]

[answer]India[/answer]
[wrong]China[/wrong]
[wrong]Switzerland[/wrong]
[wrong]Italy[/wrong]
[/quiz]

[quiz]
[question]£32,000: The Mississippi River empties into the Gulf of Mexico directly from what state?
[/question]

[answer] Louisiana[/answer]
[wrong] Mississippi[/wrong]
[wrong]Texas[/wrong]
[wrong]Alabama[/wrong]
[/quiz]

[quiz]
[question]£64,000: What currency is used in Bahrain?
[/question]

[answer]Dinar[/answer]
[wrong]Riyal[/wrong]
[wrong]Dirham[/wrong]
[wrong]Rouble[/wrong]
[/quiz]

[quiz]
[question]£125,000: Which of these cetaceans is classified as a “toothed whale”?
[/question]

[answer]Sperm whale[/answer]
[wrong]Grey whale[/wrong]
[wrong]Humpback whale[/wrong]
[wrong]Minke whale[/wrong]
[/quiz]

[quiz]
[question]£250,000: In 1718, which pirate died in battle off the coast of what is now North Carolina?
[/question]

[answer]Blackbeard[/answer]
[wrong]Calico Jack[/wrong]
[wrong]Captain Kidd[/wrong]
[wrong]Bartholomew Roberts[/wrong]
[/quiz]

[quiz]
[question]£500,000: In mythology, which creature sprang from the blood of the Medusa?
[/question]

[answer]Pegasus[/answer]
[wrong]Centaur[/wrong]
[wrong]Hydra[/wrong]
[wrong]Penguin[/wrong]
[/quiz]

[quiz]
[question]£1,000,000: Who wrote the opera ‘The Thieving Magpie’?
[/question]

[answer]Rossini[/answer]
[wrong]Verdi[/wrong]
[wrong]Donizetti[/wrong]
[wrong]Puccini[/wrong]
[/quiz]

[/quiz-wrapper]
```

### CODE REVIEW

File: projects/plugins/jetpack/modules/shortcodes/css/quiz.css [hunk 1]
```diff
@@ -54,3 +54,35 @@ div.jetpack-quiz pre {
...
+	margin-left: 12px;
+}
+
+.jetpack-quiz-option-button:hover svg {
+	background: #3c434a;
+}
+
+.jetpack-quiz-option-button svg {
+	background: #000;
+	border-radius: 50%;
```
Comment (jeherve) [inline projects/plugins/jetpack/modules/shortcodes/css/quiz.css:79]: I would suggest adding a display to ensure things will be displayed nicely in themes that already style SVGs:

```suggestion
	border-radius: 50%;
	display: inline-block;
```

**Review by jeherve — DISMISSED** [review]
I like it, that's a great addition! I only have one suggestion, otherwise I think this is good to go.

**Review by jeherve — APPROVED** [review]

**Review by jeherve — DISMISSED** [review]
Coming back to this, I'm only noticing now that we have some linting issues in there, most likely caused by the es6 code in your JavaScript additions.

```
jetpack  update/quiz-shortcode
❯ pnpm lint-file projects/plugins/jetpack/modules/shortcodes/js/quiz.js

> Jetpack_Monorepo@ lint-file /Users/jeherve/code/jetpack
> eslint --ext .js,.jsx,.cjs "projects/plugins/jetpack/modules/shortcodes/js/quiz.js"


projects/plugins/jetpack/modules/shortcodes/js/quiz.js
  66:5  error  Parsing error: Unexpected token quizWrapper

✖ 1 problem (1 error, 0 warnings)

 ERROR  Command failed with exit code 1.
```

Could you edit this so it uses es5 instead, like the rest of the shortcodes' JavaScript?

Thank you!

**Comment by Aurorum** [comment]
Thanks @jeherve, glad you like it! That issue should be addressed with the latest commit. :)

**Comment by jeherve** [comment]
> That issue should be addressed with the latest commit. :)

I'm afraid it won't let you use es6's arrow functions:

```
/home/runner/work/jetpack/jetpack/projects/plugins/jetpack/modules/shortcodes/js/quiz.js
  66:73  error  Parsing error: Unexpected token >
```

**Comment by Aurorum** [comment]
Oops, I've removed the arrow functions now - apologies, I've not been able to run the tests locally unfortunately.

**Review by samiff — COMMENTED** [review]
Hey @Aurorum Very cool enhancement you've worked on here 👍   I like that it's an optional layer over the existing quiz shortcode.

My only concern was around the use of the links:

`<a class="jetpack-quiz-option-button" data-quiz-option="next">`

For navigation, as I don't think this would be very accessible as-is:

- https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/button_role
- https://www.w3.org/TR/wai-aria-practices/examples/button/button.html

Though I'm not sure if that should be a blocker here as I was having issues with the current quiz shortcode navigating via VoiceOver on Mac (e.g. wanting to focus via keyboard and hit Enter on an answer).

Perhaps a11y could be looked at in a separate issue, @jeherve what are your thoughts here?

**Comment by jeherve** [comment]
> My only concern was around the use of the links

Oh, that's a great point, you're right. In b5ed13c731fe527426d6db8ca6baa1002c90fb30 I added some extra info to those "buttons". What do you think about it?

**Comment by samiff** [comment]
> In [b5ed13c](https://github.com/Automattic/jetpack/commit/b5ed13c731fe527426d6db8ca6baa1002c90fb30) I added some extra info to those "buttons". What do you think about it?

I think that I have a lot to learn about a11y 😄 

At least now when navigating to the buttons they are read off as "Next/Previous Quiz Button" instead of being read as an untitled image which is great.

However when the button is "clicked" (I'm using VoiceOver), it's pretty confusing because there is no focus handling going on here, so there isn't a good way to know that a new quiz has been rendered, or easily navigate to that new quiz question top.

I think this would go back to a larger a11y review though. I'm okay with the changes otherwise, so unless you think the a11y angle is blocking @jeherve we could get this merged.

**Comment by jeherve** [comment]
> I think that I have a lot to learn about a11y 😄

Same! :)

> However when the button is "clicked" (I'm using VoiceOver), it's pretty confusing because there is no focus handling going on here, so there isn't a good way to know that a new quiz has been rendered, or easily navigate to that new quiz question top.

That's a good point. Maybe we could address this by adding `.focus()` when we click and switch view to to the next quiz here?
https://github.com/Automattic/jetpack/pull/20592/files#diff-912dadc30c75b3d91095824d02ea1dda6b14d89e7b0489c168d2f1a019639e08R86

**Review by samiff — APPROVED** [review]
> Maybe we could address this by adding `.focus()` when we click and switch view to to the next quiz

I think that will help push things forward here, added in: https://github.com/Automattic/jetpack/pull/20592/commits/a761cfe85556089b1dc85e57f0877dd910ca03fa

Now when navigating between quizzes, the quiz question will be focused and read off 👍 

@Aurorum Thanks again for the contribution! This change should ship out in Jetpack 10.1 which is scheduled to release on September 7th, 2021.

**Comment by samiff** [comment]
## Internal

Diff: D65564-code
WPcom doc request: todo
JPOP doc request: todo

**Comment by Aurorum** [comment]
Thanks both for the help with this. :)

**Pull request merged**